src/databricks/labs/ucx/hive_metastore/migration_status.py

```diff
@@ -0,0 +1,129 @@
+        dst = self._index.get((schema.lower(), table.lower()))
+        if not dst or not dst.dst_table:
+            return None
+        return dst
+
+
+class MigrationStatusRefresher(CrawlerBase[MigrationStatus]):
+    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema, table_crawler: TablesCrawler):
+        super().__init__(sbe, "hive_metastore", schema, "migration_status", MigrationStatus)
+        self._ws = ws
+        self._table_crawler = table_crawler
+
+    def snapshot(self) -> Iterable[MigrationStatus]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    def index(self) -> MigrationIndex:
+        return MigrationIndex(list(self.snapshot()))
+
+    def get_seen_tables(self) -> dict[str, str]:
+        seen_tables: dict[str, str] = {}
+        for schema in self._iter_schemas():
+            try:
+                tables = self._ws.tables.list(catalog_name=schema.catalog_name, schema_name=schema.name)
+            except NotFound:
+                logger.warning(
+                    f"Schema {schema.catalog_name}.{schema.name} no longer exists. Skipping checking its migration status."
+                )
+                continue
+            for table in tables:
+                if not table.properties:
+                    continue
+                if "upgraded_from" not in table.properties:
+                    continue
+                if not table.full_name:
+                    logger.warning(f"The table {table.name} in {schema.name} has no full name")
+                    continue
+                seen_tables[table.full_name.lower()] = table.properties["upgraded_from"].lower()
+        return seen_tables
+
+    def is_migrated(self, schema: str, table: str) -> bool:
+        result = self._backend.fetch(
+            f"SHOW TBLPROPERTIES {escape_sql_identifier(schema + '.' + table)} ('upgraded_to')"
+        )
+        for _ in result:
+            logger.info(f"{schema}.{table} is set as migrated")
+            return True
+        logger.info(f"{schema}.{table} is set as not migrated")
+        return False
+
+    def _crawl(self) -> Iterable[MigrationStatus]:
+        all_tables = self._table_crawler.snapshot()
+        reverse_seen = {v: k for k, v in self.get_seen_tables().items()}
+        timestamp = datetime.datetime.now(datetime.timezone.utc).timestamp()
+        for table in all_tables:
+            src_schema = table.database.lower()
+            src_table = table.name.lower()
+            table_migration_status = MigrationStatus(
+                src_schema=src_schema,
+                src_table=src_table,
+                update_ts=str(timestamp),
+            )
+            if table.key in reverse_seen and self.is_migrated(src_schema, src_table):
+                target_table = reverse_seen[table.key]
+                if len(target_table.split(".")) == 3:
+                    table_migration_status.dst_catalog = target_table.split(".")[0]
```

Comment:
We use [`dataclasses.replace`](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) to overwrite data class instance properties

Reply:
Sure, I can do that.